traffic_ops/traffic_ops_golang/deliveryservice/keys_test.go

```diff
@@ -20,214 +20,728 @@ package deliveryservice
+		t.FailNow()
+	}
+
+	if privateKey == nil {
+		t.Error("RSA private key is nil. expect: not nil")
+		t.FailNow()
+	}
+}
+
+func TestDecodePrivateKey_PKCS1_RSA_2048(t *testing.T) {
+
+	privateKey, cleanPemPrivateKey, err := decodePrivateKey(PrivateKey_PKCS1_RSA_2048)
 
+	if err != nil {
+		t.Errorf("Unexpected result: " + err.Error())
+	}
+
+	pBlock, remain := pem.Decode([]byte(cleanPemPrivateKey))
+
+	if pBlock == nil {
+		t.Error("can't decode cleaned private key pem block")
+		t.FailNow()
+	} else if len(remain) > 0 {
+		t.Error("remaining bytes after decode > 0. expected: 0")
+		t.FailNow()
+	}
+
+	if privateKey == nil {
+		t.Error("RSA private key is nil")
+		t.FailNow()
+	}
+}
+
+func TestDecodePrivateKey_RSA_Encrypted_And_BadData(t *testing.T) {
+
+	// Expected to fail on decode of encrypted pem private key
+	privateKey, _, err := decodePrivateKey(PrivateKey_ENCRYPTED_RSA_2048)
+	if err == nil && privateKey != nil {
+		t.Error("unexpected result: decoding of encrypted private key should have returned an error")
+		t.FailNow()
+	}
+
+	// Expected to fail.
+	privateKey, _, err = decodePrivateKey(BadKeyData)
+	if err == nil && privateKey != nil {
+		t.Error("unexpected result: decoding of bad private key data should have returned an error")
+		t.FailNow()
+	}
+}
+
+func TestVerifyAndEncodeCertificate_BadData(t *testing.T) {
 	// should fail bad base64 data
-	dat, _, err := verifyCertificate(BadData, "")
+	dat, pemEncodedKey, unknownAuth, err := verifyCertKeyPair(BadCertData, BadKeyData, "")
 	if err == nil {
 		t.Errorf("Unexpected result, there should have been a base64 decoding failure")
 	}
 
-	// should fail, can't verify self signed cert against this rootCA
-	dat, _, err = verifyCertificate(SelfSignedCertOnly, rootCA)
-	if err == nil {
+	_ = dat
+	_ = pemEncodedKey
+	_ = unknownAuth
+}
+
+func TestVerifyAndEncodeCertificate_SelfSigned_NO_SKIAKI_CertKeyPair(t *testing.T) {
+
+	certChain, privateKey, unknownAuth, err := verifyCertKeyPair(SelfSignedNOSKIAKI_Certificate, SelfSignedNOSKIAKI_PrivateKey, "")
+
+	if err != nil {
+		t.Errorf("Unexpected result, a certificate verification error should have occured")
+	}
+
+	if !unknownAuth {
+		t.Errorf("Unexpected result, certificate verification should have detected unknown authority")
+	}
+
+	// Decode the clean Private Key
+	pBlock, remain := pem.Decode([]byte(privateKey))
+
+	if pBlock == nil {
+		t.Error("unexpected result: can't decode cleaned private key pem block")
+		t.FailNow()
+	} else if len(remain) > 0 {
+		t.Error("unexpected result: remaining bytes after decode > 0. expected: 0")
+		t.FailNow()
+	}
+
+	if len(certChain) == 0 {
+		t.Error("unexpected: certchain should not empty")
+	}
+}
+
+func TestVerifyAndEncodeCertificate_SelfSignedCertKeyPair(t *testing.T) {
+
+	certChain, privateKey, unknownAuth, err := verifyCertKeyPair(SelfSigned_Certificate, SelfSigned_PrivateKey, "")
+
+	if err != nil {
 		t.Errorf("Unexpected result, a certificate verification error should have occured")
```

Comment:
Is this the right error message in this case? `err != nil` seems like there _was_ a verification error.